server/src/main/java/org/opensearch/index/ShardIndexingPressureMemoryManager.java

```diff
@@ -0,0 +1,458 @@
+                                     BiPredicate<Long, Long> updateShardLimitPredicate, long shardBaseLimit) {
+
+        long currentShardLimit;
+        long newShardLimit;
+        do {
+            currentShardLimit = shardLimitSupplier.getAsLong();
+            long shardCurrentBytes = shardCurrentBytesSupplier.getAsLong();
+            newShardLimit = Math.max((long) (shardCurrentBytes / this.optimalOperatingFactor), shardBaseLimit);
+
+            if (((double)shardCurrentBytes / currentShardLimit) > this.lowerOperatingFactor) {
+                logger.debug("Shard Limits Already Decreased [" +
+                        "shard_detail=[{}][{}], " + "current_shard_limit_bytes={}, " +
+                        "expected_shard_limit_bytes={}]",
+                    shardId.getIndexName(), shardId.id(), currentShardLimit, newShardLimit);
+                return;
+            }
+        } while(!updateShardLimitPredicate.test(currentShardLimit,newShardLimit));
+
+        logger.debug("Decreased Shard Limit [shard_detail=[{}][{}], " +
+                "old_shard_limit_bytes={}, new_shard_limit_bytes={}]",
+            shardId.getIndexName(), shardId.id(), currentShardLimit, newShardLimit);
+    }
+
+    /**
+     * This evaluation returns true if throughput of last N request divided by the total lifetime requests throughput is greater than
+     * the degradation limits threshold.
+     */
+    private boolean evaluateThroughputDegradationLimitsBreached(PerformanceTracker performanceTracker, StatsTracker statsTracker,
+                                                                double degradationLimits) {
+        double throughputMovingAverage =  Double.longBitsToDouble(performanceTracker.getThroughputMovingAverage());
+        long throughputMovingQueueSize = performanceTracker.getThroughputMovingQueueSize();
+        double throughputHistoricalAverage = (double)statsTracker.getTotalBytes() / performanceTracker.getLatencyInMillis();
+        return throughputMovingAverage > 0 && throughputMovingQueueSize >= this.shardIndexingPressureSettings.getRequestSizeWindow() &&
+            throughputHistoricalAverage / throughputMovingAverage > degradationLimits;
+    }
+
+    /**
+     * This evaluation returns true if the difference in the current timestamp and last successful request timestamp is greater than
+     * the successful request elapsed-timeout threshold, and the total number of outstanding requests is greater than
+     * the maximum outstanding request-count threshold.
+     */
+    private boolean evaluateLastSuccessfulRequestDurationLimitsBreached(PerformanceTracker performanceTracker, long requestStartTime) {
+        return (performanceTracker.getLastSuccessfulRequestTimestamp() > 0) &&
+            (requestStartTime - performanceTracker.getLastSuccessfulRequestTimestamp()) > this.successfulRequestElapsedTimeout &&
+            performanceTracker.getTotalOutstandingRequests() > this.maxOutstandingRequests;
+    }
```

Comment:
Lets make them package private to test the logic out

Reply:
Effects of these methods are already tested under tests for the public methods of the class, such as tests with suffixes:
```
SoftLimitBreachedAndLastSuccessfulRequestLimitRejections
SoftLimitBreachedAndLessOutstandingRequestsAndNoLastSuccessfulRequestLimitRejections
SoftLimitBreachedAndThroughputDegradationLimitRejections
SoftLimitBreachedAndMovingAverageQueueNotBuildUpAndNoThroughputDegradationLimitRejections
SoftLimitBreachedAndNoSecondaryParameterBreachedAndNodeLevelRejections
```